### PR TITLE
[WIP][SPARK-31168][BUILD][test-maven] Upgrade Scala to 2.12.11

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR(SPARK-31168) upgrade Scala to **2.12.11**.

Scala 2.12.11 release notes:
**Highlights**
- Performance improvements in the collections library
- Performance improvements in the compiler
- Improvements to `-Yrepl-class-based`, an alternative internal REPL encoding that avoids deadlocks
- A new `-Yrepl-use-magic-imports` flag that avoids deep class nesting in the REPL, which can lead to deteriorating performance in long sessions
- Fix some toX methods that could expose the underlying mutability of a ListBuffer-generated collection

**JDK 9+ support**
- ASM was upgraded to 7.3.1, allowing the optimizer to run on JDK 13+
- `:javap` in the REPL now works on JDK 9+

### Why are the changes needed?
Improvement performance.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Exist test.

### CODE REVIEW

File: docs/_config.yml [hunk 1]
```diff
@@ -17,7 +17,7 @@ include:
 SPARK_VERSION: 3.1.0-SNAPSHOT
 SPARK_VERSION_SHORT: 3.1.0
 SCALA_BINARY_VERSION: "2.12"
-SCALA_VERSION: "2.12.10"
+SCALA_VERSION: "2.12.1"
```
Comment (srowen) [inline docs/_config.yml]: Typo here

**Review by srowen — COMMENTED** [review]
I think the important thing will be to make sure the items tested at https://github.com/apache/spark/pull/27681 work

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
It seems that we need to make a PR to `genjavadoc`.
- https://github.com/lightbend/genjavadoc/blob/master/build.sbt#L51

**Comment by maropu** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by maropu** [comment]
(Does this upgrade mak the testing time much longer..?)

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by wangyum** [comment]
Retest this please.

**Comment by maropu** [comment]
retest this please

**Comment by wangyum** [comment]
It seems 2.12.11 breaks generate Java doc:  https://github.com/scala/bug/issues/11917

**Comment by wangyum** [comment]
Close it due to this issue: https://github.com/scala/bug/issues/11917

**Comment by dongjoon-hyun** [comment]
Thank you, @wangyum .
I updated SPARK-31168 from `Upgrade Scala to 2.12.11` to `Upgrade Scala to 2.12.12`.

**Pull request closed without merging**